compiler/cpp/src/thrift/generate/t_js_generator.cc

```diff
@@ -1580,10 +1580,10 @@ void t_js_generator::generate_service_client(t_service* tservice) {
       } else {
         f_service_ << indent() << "if (callback) {" << endl;
         f_service_ << indent() << "  var self = this;" << endl;
-        f_service_ << indent() << "  this.output.getTransport().flush(true, function() {" << endl;
+        f_service_ << indent() << "  this.output.getTransport().flush(true, function(xhrError) {" << endl;
         f_service_ << indent() << "    var result = null;" << endl;
         f_service_ << indent() << "    try {" << endl;
-        f_service_ << indent() << "      result = self.recv_" << funname << "();" << endl;
+        f_service_ << indent() << "      result = self.recv_" << funname << "() || xhrError;" << endl;
```

Comment:
The jQuery generated code doesn't use the native browser XHR objects directly, it relies on jQuery wrapping it. Because of that, the callbacks used with the jQuery code do have the errors passed back to them. This is clearly a gap between the jQuery and vanilla JS options.

I think if the `recv_*()` successfully returns something, and it often does, it should supercede whatever error we've encountered. I want to leave as small a footprint as possible on this code in case we need to upgrade Thrift again and the behavior of what the `recv_*()` functions can return changes. If that function returns a value AND we see an error, changes are Thrift is purposefully handling that case the "right" way, and we shouldn't get in the way of that.

Ultimately the real way it should be handled is the `function(error, result) { ... }` callback argument convention NodeJS uses. The `error` argument is always `null` unless it's an error case.

Yes, this would definitely be a place that would be touched when changing to Promises.